src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs

```diff
@@ -383,49 +377,68 @@ protected override void Dispose(bool disposing)
             }
         }
 
-        private readonly struct RequestRegistration
+        private sealed class RequestRegistration : TaskCompletionSource<RateLimitLease>
         {
-            public RequestRegistration(int requestedCount, TaskCompletionSource<RateLimitLease> tcs,
-                CancellationTokenRegistration cancellationTokenRegistration)
-            {
-                Count = requestedCount;
-                // Perf: Use AsyncOperation<TResult> instead
-                Tcs = tcs;
-                CancellationTokenRegistration = cancellationTokenRegistration;
-            }
+            private readonly CancellationToken _cancellationToken;
+            private CancellationTokenRegistration _cancellationTokenRegistration;
 
-            public int Count { get; }
+            // this field is used only by the disposal mechanics and never shared between threads
+            private RequestRegistration? _next;
 
-            public TaskCompletionSource<RateLimitLease> Tcs { get; }
+            public RequestRegistration(int permitCount, ConcurrencyLimiter limiter, CancellationToken cancellationToken)
+                : base(limiter, TaskCreationOptions.RunContinuationsAsynchronously)
+            {
+                Count = permitCount;
+                _cancellationToken = cancellationToken;
 
-            public CancellationTokenRegistration CancellationTokenRegistration { get; }
-        }
+                // RequestRegistration objects are created while the limiter lock is held
+                // if cancellationToken fires before or while the lock is held, UnsafeRegister
+                // is going to invoke the callback synchronously, but this does not create
+                // a deadlock because lock are reentrant
+                if (cancellationToken.CanBeCanceled)
+#if NETCOREAPP || NETSTANDARD2_1
```

Comment:
```suggestion
#if NETCOREAPP || NETSTANDARD2_1_OR_GREATER
```

Reply:
Maybe just keep it as is. Most libraries in this repo use `NETSTANDARD2_1`, and `NETSTANDARD2_1_OR_GREATER` is supported starting with the 5.0 SDK.

Reply:
I updated before reading your comment 🤣 5.0 has been EOL for more than a year too.